File: backend/src/routes/download.ts
import { Router, Request, Response, NextFunction } from 'express';
import archiver from 'archiver';
import { Readable } from 'stream';
import * as ShareService from '../services/ShareService';
import * as StorageService from '../services/StorageService';
import { normalizeUID, isValidUID } from '../services/UIDService';
import { validateSessionToken } from '../services/PasswordService';
import { retrievalLimiter } from '../middleware/rateLimiter';
import { createError } from '../middleware/errorHandler';

const router = Router();

// ─── GET /api/download/:uid/all — ZIP stream of all files ───
router.get(
  '/:uid/all',
  retrievalLimiter,
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const uid = normalizeUID(req.params.uid);
      if (!isValidUID(uid)) {
        throw createError('Invalid UID format', 400);
      }

      // Get share metadata to check privacy
      const share = await ShareService.getShareByUID(uid);
      if (!share) {
        throw createError('Share not found or has expired', 404);
      }

      // Auth check for private shares
      if (share.isPrivate) {
        const token = req.headers['x-session-token'] as string;
        if (!token || !validateSessionToken(token, uid)) {
          throw createError('Authentication required', 401);
        }
      }

      // Get all files for this share
      const files = await ShareService.getShareFiles(uid);
      if (!files || files.length === 0) {
        throw createError('No files found for this share', 404);
      }

      // Set up ZIP streaming response
      res.setHeader('Content-Type', 'application/zip');
      res.setHeader('Content-Disposition', `attachment; filename="sharenova-${uid}.zip"`);

      const archive = archiver('zip', { zlib: { level: 5 } });
      archive.pipe(res);

      // Stream each file from R2 into the ZIP
      for (const file of files) {
        const stream = await StorageService.getFileStream(file.storageKey);
        if (stream) {
          archive.append(stream as Readable, { name: file.filename });
        }
      }

      await archive.finalize();
    } catch (error) {
      next(error);
    }
  }
);

export default router;

File: backend/app/routers/download.py
"""
download.py — ZIP streaming download of all files in a share.
Replaces: backend/src/routes/download.ts (archiver npm → Python zipfile)
"""

import io
import zipfile

from fastapi import APIRouter, Depends, Request, Header
from fastapi.responses import JSONResponse, StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services import share_service, storage_service, password_service
from app.services.uid_service import normalize_uid, is_valid_uid
from app.middleware.rate_limiter import limiter, RETRIEVAL_LIMIT

router = APIRouter(prefix="/api/download")


@router.get("/{uid}/all")
@limiter.limit(RETRIEVAL_LIMIT)
async def download_all_as_zip(
    request: Request,
    uid: str,
    db: AsyncSession = Depends(get_db),
    x_session_token: str | None = Header(default=None, alias="X-Session-Token"),
):
    clean_uid = normalize_uid(uid)
    if not is_valid_uid(clean_uid):
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Invalid UID format"},
        )

    # Get share metadata to check privacy
    share = await share_service.get_share_by_uid(db, clean_uid)
    if not share:
        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "Share not found or has expired"},
        )

    # Auth check for private shares
    if share["isPrivate"]:
        if not x_session_token or not password_service.validate_session_token(
            x_session_token, clean_uid
        ):
            return JSONResponse(
                status_code=401,
                content={"success": False, "error": "Authentication required"},
            )

    # Get all files for this share
    files = await share_service.get_share_files(db, clean_uid)
    if not files:
        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "No files found for this share"},
        )

    # Build ZIP in a streaming buffer
    async def generate_zip():
        buf = io.BytesIO()
        with zipfile.ZipFile(buf, "w", zipfile.ZIP_DEFLATED, compresslevel=5) as zf:
            for f in files:
                file_bytes = await storage_service.get_file_bytes(f["storage_key"])
                if file_bytes:
                    zf.writestr(f["filename"], file_bytes)
        buf.seek(0)
        yield buf.read()

    return StreamingResponse(
        generate_zip(),
        media_type="application/zip",
        headers={
            "Content-Disposition": f'attachment; filename="sharenova-{clean_uid}.zip"'
        },
    )
